milli/src/update/word_prefix_pair_proximity_docids.rs

```diff
@@ -64,44 +77,47 @@ impl<'t, 'u, 'i> WordPrefixPairProximityDocids<'t, 'u, 'i> {
         );
 
         let prefix_fst = self.index.words_prefixes_fst(self.wtxn)?;
-        let prefix_fst_keys = prefix_fst.into_stream().into_bytes();
-        let prefix_fst_keys: Vec<_> = prefix_fst_keys
-            .as_slice()
-            .linear_group_by_key(|x| std::str::from_utf8(&x).unwrap().chars().nth(0).unwrap())
-            .collect();
+        let prefix_fst_keys = prefix_fst.into_stream().into_strs()?;
+        let prefix_fst_keys: Vec<_> =
+            prefix_fst_keys.as_slice().linear_group_by_key(|x| x.chars().nth(0).unwrap()).collect();
 
         let mut db =
             self.index.word_pair_proximity_docids.remap_data_type::<ByteSlice>().iter(self.wtxn)?;
 
         let mut buffer = Vec::new();
-        let mut current_prefixes: Option<&&[Vec<u8>]> = None;
+        let mut current_prefixes: Option<&&[String]> = None;
         let mut prefixes_cache = HashMap::new();
         while let Some(((w1, w2, prox), data)) = db.next().transpose()? {
+            if prox > self.max_proximity {
+                continue;
+            }
+
             current_prefixes = match current_prefixes.take() {
-                Some(prefixes) if w2.as_bytes().starts_with(&prefixes[0]) => Some(prefixes),
+                Some(prefixes) if w2.starts_with(&prefixes[0]) => Some(prefixes),
                 _otherwise => {
                     write_prefixes_in_sorter(
                         &mut prefixes_cache,
                         &mut word_prefix_pair_proximity_docids_sorter,
-                        self.threshold,
                     )?;
-                    prefix_fst_keys.iter().find(|prefixes| w2.as_bytes().starts_with(&prefixes[0]))
+                    prefix_fst_keys.iter().find(|prefixes| w2.starts_with(&prefixes[0]))
                 }
             };
 
             if let Some(prefixes) = current_prefixes {
                 buffer.clear();
                 buffer.extend_from_slice(w1.as_bytes());
                 buffer.push(0);
-                for prefix in prefixes.iter().filter(|prefix| w2.as_bytes().starts_with(prefix)) {
-                    buffer.truncate(w1.len() + 1);
-                    buffer.extend_from_slice(prefix);
-                    buffer.push(prox);
-
-                    match prefixes_cache.get_mut(&buffer) {
-                        Some(value) => value.push(data),
-                        None => {
-                            prefixes_cache.insert(buffer.clone(), vec![data]);
+                for prefix in prefixes.iter() {
+                    if prefix.len() <= self.max_prefix_length && w2.starts_with(prefix) {
```

Comment:
the condition `prefix.len() <= self.max_prefix_length` has probably no cost compared to the rest, but, shouldn't we prefilter prefixes coming from the fst before beginning the computation?

Reply:
Do you mean that it would be more logical to change the order of both conditions in this if?

Reply:
no, this condition could be removed from this line being done before the double loop line 88:
```rust
// filter here on prefix length.
while let Some(((w1, w2, prox), data)) = db.next().transpose()? {
                // [...]
                for prefix in prefixes.iter() {
                    // instead of here.
                    if prefix.len() <= self.max_prefix_length && w2.starts_with(prefix) {
                        // [...]
                    }
               }
}
```

Reply:
I don't know if it is possible, but, can you filter directly in the FST the prefixes that have a length under `self.max_prefix_length`

Reply:
It is possible to filter out the prefixes by their lengths when we store them in the `prefix_fst_keys` vec but it would imply that we create a sub-vec with all of the filtered values. Now we just have a simple slice that we receive from `slice_group_by`. I prefer to keep it like that for now.